salvo/src/lib/run_benchmark.py

```diff
@@ -0,0 +1,324 @@
+"""
+General benchmark wrapper that validates that the
+job control contains all dat required for each known
+benchmark
+"""
+import logging
+import os
+from typing import (List, Set)
+
+from src.lib.benchmark import fully_dockerized_benchmark as fulldocker
+from src.lib.benchmark import scavenging_benchmark as scavenging
+from src.lib.benchmark import base_benchmark
+from src.lib.docker_management import (docker_image, docker_image_builder)
+from src.lib import source_manager
+
+import api.control_pb2 as proto_control
+import api.source_pb2 as proto_source
+import api.image_pb2 as proto_image
+
+log = logging.getLogger(__name__)
+
+
+class BenchmarkRunnerError(Exception):
+  """An error raised if if an unrecoverable condition arises when executing
+     a benchmark.
+  """
+
+class BenchmarkRunner(object):
+  """This class contains the logic to validate input artifacts and
+     perform a benchmark.
+  """
+
+  def __init__(self, control: proto_control.JobControl) -> None:
+    """Initialize the benchmark object.
+
+    Perform class member initialization and instantiate the underlying
+    object actually performing the test.
+
+    Args:
+      control: The Job Control object dictating the parameters governing the
+      benchmark
+
+    Returns:
+      None
+    """
+    self._control = control
+    self._source_manager = source_manager.SourceManager(self._control)
+
+    self._test = []
+    self._setup_test()
+
+  def _setup_test(self) -> None:
+    """Create the object performing the desired benchmark.
+
+    Instantiate the object performing the actual test and setup the control
+    objects for each test invocation.
+
+    Raises:
+      NotImplementedError: for tests and/or modes that are not yet implemented.
+    """
+
+    current_benchmark_name = "Unspecified Benchmark"
+
+    if self._control.scavenging_benchmark:
+      current_benchmark_name = "Scavenging Benchmark"
+      job_control_list = self.generate_job_control_for_envoy_images()
+
+      for job_control in job_control_list:
+        benchmark = scavenging.Benchmark(job_control, current_benchmark_name)
+        self._test.append(benchmark)
+
+    elif self._control.dockerized_benchmark:
+      current_benchmark_name = "Fully Dockerized Benchmark"
+      job_control_list = self.generate_job_control_for_envoy_images()
+
+      for job_control in job_control_list:
+        benchmark = fulldocker.Benchmark(job_control, current_benchmark_name)
+        self._test.append(benchmark)
+
+    elif self._control.binary_benchmark:
+      current_benchmark_name = "Binary Benchmark"
+      job_control_list = []
+
+    if not self._test:
+      raise NotImplementedError(f"No [{current_benchmark_name}] defined")
+
+  def generate_job_control_for_envoy_images(self) \
```

Comment:
This class has a few more public methods (other than execute). Assuming they do need to be public, can we also add unit tests for them, so that every publicly available method has its own set of tests that verify it?

Reply:
I can make them private.  They aren't really useful when called externally to the module